### PR TITLE
Improve torch.fx support to trace non torch objects

### PR DESCRIPTION
**Pull request by thomasw21**
# What does this PR do?

Improve `torch.fx` support to be able to track non tensor/parameter/module attributes. For example the new tracing mechanism allows to trace `config`, `num_heads` etc ... attributes that are important if one wants to build transformations on top.

Caveat: tracing such objects prevents `torch.jit.script` to work correctly due to limitations on their end. Consequently we gate the feature behind `trace_non_torch_object` flag upon tracing which defaults to False. (We could default to True, but that's a breaking change)

```
E   AssertionError: Could not TorchScript the traced model: 
E   Module 'GraphModule' has no attribute 'config' (This attribute exists on the Python module, but we failed to convert Python type: 'transformers.models.gpt2.configuration_gpt2.GPT2Config' to a TorchScript type. Only tensors and (possibly nested) tuples of tensors, lists, or dictsare supported as inputs or outputs of traced functions, but instead got value of type GPT2Config.. Its type was inferred; try adding a type annotation for the attribute.):
E     File "<eval_with_key>.0", line 5
E   def forward(self, input_ids : torch.Tensor, token_type_ids : torch.Tensor):
E       config = self.config
E                ~~~~~~~~~~~ <--- HERE
E       config_1 = self.config
E       config_2 = self.config
```

### CODE REVIEW

File: src/transformers/utils/fx.py [hunk 1]
```diff
@@ -538,11 +538,27 @@ class HFProxy(Proxy):
...
+                f" {metadata}"
+            )
         self._metadata = metadata
 
     @property
     def shape(self):
-        return self.tracer.create_proxy("call_method", "size", (self,), {})
+        return self.size()
+
+    def size(self, dim=None):
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:552]: Question: can this be removed because `__iter__` solves also the issue?
If so, let's keep `shape` and remove `size`?

File: src/transformers/utils/fx.py [hunk 2]
```diff
@@ -566,9 +582,25 @@ def __bool__(self):
...
+                    proxy.install_metadata(self._metadata[i])
+                return proxies.__iter__()
+            else:
+                return self._metadata.__iter__()
+        return super().__iter__()
+
     def __getattr__(self, k):
         if k == "_metadata":
-            return self.__getattribute__(k)
+            return super().__getattribute__(k)
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:602]: Why?

File: src/transformers/utils/fx.py [hunk 3]
```diff
@@ -582,8 +614,31 @@ def __contains__(self, key):
...
+
+        # Without `_metadata` we create a proxy to get such an element.
+        return self.tracer.create_proxy(
+            "get_attr", item, (), {}, proxy_factory_fn=lambda node: HFModelAttribute(node, self.tracer)
+        )
+
+
 class HFAttribute(HFProxy):
     def __init__(self, root, attr: str):
+        assert attr != "_metadata", "`_metadata` is an invalid attribute for a proxy"
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:641]: No assert in `transformers`
```suggestion
         if attr == "_metadata":
             raise AttributeError("_metadata is an invalid attribute for a proxy")
```

File: src/transformers/utils/fx.py [hunk 4]
```diff
@@ -926,6 +1099,9 @@ def trace(
             complete_concrete_args_with_inputs_not_in_dummy_inputs (`bool`, *optional*, defaults to `True`):
                 If `True`, and `dummy_inputs` is specified, every argument that `root` can take that is not in
                 `dummy_inputs` and not in `concrete_args` will be added to `concrete_args`, otherwise does nothing.
+            trace_non_torch_object (`bool`, *optional*, defaults to `False`):
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:1102]: I think this should be a `HFTracer` class attribute to match what is being done in PyTorch: https://github.com/pytorch/pytorch/blob/master/torch/fx/proxy.py#L26

File: src/transformers/utils/fx.py [hunk 5]
```diff
@@ -947,29 +1123,23 @@ def trace(
...
-        for input_name in input_names:
-            if input_name in inputs:
-                continue
+        input_names_to_generate_dummies = set(input_name for input_name in input_names if input_name not in inputs)
+
+        # sanity check
+        for input_name in input_names_to_generate_dummies:
             # We enforce that root must either be a PreTrainedModel or deserialized from a serialized traced model to
             # be able to use HFTracer._generate_dummy_input.
             if isinstance(root, PreTrainedModel) or type(root).__qualname__.startswith("_deserialize_graph_module"):
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:1133]: Maybe reformulate as `if not ...` instead of `if / else`.

File: src/transformers/utils/fx.py [hunk 6]
```diff
@@ -646,6 +715,75 @@ def _generate_random_int(low: int = 10, high: int = 20, forbidden_values: Option
     return value
 
 
+class Patch:
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:718]: I think it can be good to document the expected structure of `patched_methods`

File: src/transformers/utils/fx.py [hunk 7]
```diff
@@ -646,6 +715,75 @@ def _generate_random_int(low: int = 10, high: int = 20, forbidden_values: Option
...
+            for __o, name, orig, wrapper in original_methods:
+                setattr(__o, name, wrapper)
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        for _, original_methods in self.patched_methods.items():
+            for __o, name, orig, wrapper in original_methods:
+                setattr(__o, name, orig)
+
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:733]: Should this return `self` just as `__enter__`? Not very familiar with context managers...

File: src/transformers/utils/fx.py [hunk 8]
```diff
@@ -978,24 +1148,39 @@ def trace(
             if param.kind == inspect.Parameter.VAR_KEYWORD and param.name not in input_names:
                 concrete_metas[f"**{param.name}"] = {}
         self.meta_args = concrete_metas
-        self.patched_torch_methods = {
-            target: _gen_constructor_wrapper(getattr(torch, target)) for target in self._TORCH_METHODS_TO_PATCH
+        self.orig_fns = set(getattr(torch, target) for target in self._TORCH_METHODS_TO_PATCH)
+
+        patched_torch_methods = {
+            f"torch.{target}": [
+                (torch, target, getattr(torch, target), _gen_constructor_wrapper(getattr(torch, target)))
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:1155]: Isn't there redundancy in specifying the module, the attribute name, and the attribute?
With the (module, attr_name) we can always do `getattr(module, attr_name)` internally right?

File: src/transformers/utils/fx.py [hunk 9]
```diff
@@ -634,7 +703,7 @@ def check_has_proxy(v):
         else:
             return target(*args, **kwargs)
 
-    return wrapper, target
+    return wrapper
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:706]: I would rather keep the previous implementation as this is used in other packages. Maybe we should make this function public in that aspect.

File: src/transformers/utils/fx.py [hunk 10]
```diff
@@ -616,6 +671,20 @@ def _proxies_to_metas(v):
     return v
 
 
+def _patch_fn_for_proxy(fn, tracer):
+    def _apply_on_proxy(x):
+        if isinstance(x, torch.fx.Proxy):
+            if not (isinstance(x, HFProxy) and hasattr(x, "_metadata")):
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:677]: Maybe we can test that at the top of `_apply_on_proxy` and remove the first test.

File: src/transformers/utils/fx.py [hunk 11]
```diff
@@ -616,6 +671,20 @@ def _proxies_to_metas(v):
     return v
 
 
+def _patch_fn_for_proxy(fn, tracer):
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:674]: Why is this needed compared to what is being done in `HFTracer.create_proxy`?

File: src/transformers/utils/fx.py [hunk 12]
```diff
@@ -646,6 +715,75 @@ def _generate_random_int(low: int = 10, high: int = 20, forbidden_values: Option
...
+        # Already handled internally by `torch.fx`
+        if isinstance(attribute, (nn.Module, torch.Tensor, nn.Parameter)):
+            return attribute
+
+        # We can just get the methods instead of creating a proxy for getting it.
+        if callable(attribute):
+            return attribute
+
+        # If we're in a leaf module, let torch handle it:
+        with Patch(
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:757]: Why do we need to patch stuff to call `is_leaf_module`?

File: src/transformers/utils/fx.py [hunk 13]
```diff
@@ -646,6 +715,75 @@ def _generate_random_int(low: int = 10, high: int = 20, forbidden_values: Option
...
+            with Patch(
+                {
+                    "nn.Module.__getattribute__": [
+                        (nn.Module, "__getattribute__", nn.Module.__getattribute__, orig_get_attribute)
+                    ]
+                }
+            ):
+                prefix = tracer.path_of_module(module)
+                name = f"{prefix}.{name}" if prefix != "" else name
+                proxy = tracer.create_proxy("get_attr", name, (name,), {}, proxy_factory_fn=lambda node: HFModelAttribute(node, tracer))  # type: ignore[arg-type]
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:778]: Why do we need to specify the name in args?

File: src/transformers/utils/fx.py [hunk 14]
```diff
@@ -646,6 +715,75 @@ def _generate_random_int(low: int = 10, high: int = 20, forbidden_values: Option
...
+
+
+def create_nn_module_getattribute_wrapper(tracer):
+    torch_module_dict = {
+        **nn.Module.__dict__,
+        **nn.Module().__dict__,
+    }
+    orig_get_attribute = nn.Module.__getattribute__
+
+    def get_weird_attribute(module, name):
```
Comment (michaelbenayoun) [inline src/transformers/utils/fx.py:742]: Maybe call this `get_attribute_wrapper`?

File: src/transformers/utils/fx.py (hunk 1 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py]: Excellent this works.

File: src/transformers/utils/fx.py (hunk 7 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py:724]: Well technically I don't need to return `self`, it's just if you want to access some values from the patch:
```
with Patch() as patch:
   blablabla
```

File: src/transformers/utils/fx.py (hunk 8 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py]: Yep you're right! This is done

File: src/transformers/utils/fx.py (hunk 9 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py:706]: hum ... this function is prefixed with `_` ... When you say other packages do you know which one? Seems like you could very easily make the change?

File: src/transformers/utils/fx.py (hunk 11 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py:677]: Discussed offline

File: src/transformers/utils/fx.py (hunk 12 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py]: Essentially it would lead to some infinite loops due to me overriding `nn.Module.__getattribute__` so I just set the original function back in place, run the test, and the put back the patched version.

File: src/transformers/utils/fx.py (hunk 10 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py:680]: Discussed offline, there's three outcomes, so necessarily you need at least two if statements.

File: src/transformers/utils/fx.py (hunk 4 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py:1098]: I'm not sure what's the benefit of having it a class attribute?

File: src/transformers/utils/fx.py (hunk 2 above)
Comment (thomasw21) [reply src/transformers/utils/fx.py]: Actually I think this test was useless, since `__getattribute__` would catch it first.

File: src/transformers/models/bloom/modeling_bloom.py [hunk 15]
```diff
@@ -64,9 +64,7 @@ def _make_causal_mask(
     # ONNX doesn't support `torch.Tensor.triu` properly, thus we use this workaround
     seq_ids = torch.arange(target_length, device=device)
     mask[:, past_key_values_length:] = seq_ids[:, None] < seq_ids[None, :]
-
-    if past_key_values_length > 0:
-        mask[:, :past_key_values_length] = False
+    mask[:, :past_key_values_length] = False
```
Comment (sgugger) [inline src/transformers/models/bloom/modeling_bloom.py:67]: Can we focus this PR on one thing only, please? The changes to the BLOOM modeling file have no link with the title of the PR.

File: src/transformers/utils/fx.py [hunk 16]
```diff
@@ -13,24 +13,24 @@
...
+from typing import Any, Callable, Dict, Iterable, List, Optional, Type, Union
 
 import torch
 from packaging import version
 from torch import nn
 from torch.fx import Graph, GraphModule, Proxy, Tracer
 from torch.fx.proxy import ParameterProxy
 
-from .. import PretrainedConfig, PreTrainedModel, logging
+from .. import BloomForCausalLM, PretrainedConfig, PreTrainedModel, logging
```
Comment (sgugger) [inline src/transformers/utils/fx.py:33]: We can't have a dependency to a specific model in this module, it's going to very easily add cyclical imports. You can use the model class name instead of the object, like all other tests in this file.

File: src/transformers/utils/fx.py [hunk 17]
```diff
@@ -678,114 +790,132 @@ def _generate_dummy_input(
...
-            seq_length = _generate_random_int(low=10000, high=20000)
-            inputs_dict[input_name] = torch.zeros(batch_size, seq_length, dtype=torch.float, device=device)
-        elif "mask" in input_name or "ids" in input_name:
-            inputs_dict[input_name] = torch.zeros(shape, dtype=torch.long, device=device)
-        else:
-            shape_with_hidden_size = shape + [model.config.hidden_size]
-            inputs_dict[input_name] = torch.zeros(shape_with_hidden_size, dtype=torch.float, device=device)
+                    inputs_dict[input_name] = torch.zeros(shape, dtype=torch.long, device=device)
+            elif "past_key_values" == input_name:
+                if isinstance(model, BloomForCausalLM):
```
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/utils/fx.py:927]: Please use a check on the model name, like all other conditions in this file. Also, we can't start having specific code for specific models in this module or it will quickly become un-maintainable.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! Left a few comments.
